Draft: Initialize Garmin feature branch incl. Mock Garmin watch for Sim

mountrcg opened the pull request:
The PR aims to 

1. make the Trio-Garmin watchface configurable from Trio on the phone.
2. implement proper sensitivityRate for Trio users
3. provide support for the popular AAPS watchface by Swissalpine
 - also provide structures to setup other watchfaces and datafields
4. provide data queueing that supports the unique and seldomly used  Garmin message queue
5. remedy the issues of watchfaces becoming un-responsive after another watchface has been used, requiering the user to re-install the Trio watchface
 - this is due to Trio still providing messages after the watchface has been completely de-activated and some severe system disadvantages of the Garmin message queue

Above the hood only the Garmin configuration view has been expanded to support different watchfaces, data type selection and switching off the messaging queue for the watchface if it is not being used. Comprehensive Help sheets have been created to explain the mechanisms.

| App config | Watchface hint |Data queueing hint |
| --- | --- | --- |
|  <img width="1206" height="2622" alt="Simulator Screenshot - Dev 16 Pro - 2025-10-30 at 12 27 26" src="https://github.com/user-attachments/assets/07be8e3a-0ed0-4a1f-99e3-a74b940b429b" /> | <img width="1206" height="2622" alt="Simulator Screenshot - Dev 16 Pro - 2025-10-17 at 18 09 32" src="https://github.com/user-attachments/assets/9f1c8046-ab18-4af2-b1c2-eb93c1bd17db" /> | <img width="1206" height="2622" alt="Simulator Screenshot - Dev 16 Pro - 2025-10-17 at 18 09 38" src="https://github.com/user-attachments/assets/7f9bdb5e-3e2f-4257-9598-320e489de955" /> |

Also attached the workflow with Garmin device enabled and without to seperate Device connection setup from Garmin App settings.:

https://github.com/user-attachments/assets/8b4ea5cc-f25b-4f71-adc9-6c2f3dd42544

Testing can only be done with real Garmin watches and appropriate prg-files sideloaded onto them.

The watchface & datafield has been re-worked to work with the config options from Trio, spacing for all models has been setup dynamically. The watchface can be found in a[ newly setup repository](https://github.com/mountrcg/Trio-GarminWatchface). Installables are in the [release](https://github.com/mountrcg/Trio-GarminWatchface/releases) section.  Happy to transfer that to nightscout if deemed appropriate.
| Watchface | Datafield |
| --- | --- |
|  <img src="https://github.com/user-attachments/assets/b5c309da-1e4d-493c-8e69-ae5937b9949f" width="400"> | <img width="300" alt="Screenshot 2025-10-15 at 18 00 59" src="https://github.com/user-attachments/assets/722c6286-9126-40be-a588-ac82b99bd960" /> |
| <img src="https://github.com/user-attachments/assets/a8f46e14-08a3-4815-b25a-0ffc34931d08" width="300"> | <img width="300" alt="Screenshot 2025-10-15 at 18 00 16" src="https://github.com/user-attachments/assets/6f65260f-fb72-44f8-a162-31651ab10a6d" /> |

As one could also use the popular Swissalpine xDrip+/Spike watchface, I will make the derivative for Trio available in the release section as prg's only, once Swissalpine agrees. It works quite well!

<img width="300" alt="Screenshot 2025-10-27 at 11 20 10" src="https://github.com/user-attachments/assets/0c1af187-4d44-455b-baed-dcb163d5d3a9" />


Of course one could always use the original Swissalpine watchface/datafield with nightscout, but it would be an online only solution needing LTE/Wifi, Nightscout site and Garmin Connect running on phone. Fine in most situations. 
Here I strive for true offline capabilities, works even without Garmin Connect running.

mountrcg committed:
Initial Garmin functionality changes

*make the Trio-Garmin watchface configurable from Trio on the phone.
* implement proper sensitivityRate for Trio users
* provide support for the popular AAPS watchface by Swissalpine
  * also provide structures to setup other watchfaces and datafields
* provide data queueing that supports the unique and seldomly used Garmin message queue
* remedy the issues of watchfaces becoming un-responsive after another watchface has been used, requiering the user to re-install the Trio watchface
  * This is due to Trio still providing messages after the watchface has been completely de-activated and some severe system disadvantages of the Garmin message queue

Above the hood only the Garmin configuration view has been expanded to support different watchfaces, data type selection and switching off the messaging queue for the watchface if it is not being used. Comprehensive Help sheets have been created to explain the mechanisms.
seperate watch settings from device list

customize navigation to skip device list

add a mock garmin device in simulator

doc

common data structure


fix unnecessary number conversions

harmonize guards and default values across watch states

swissalpine datafield uuid

common data structure from AAPS

mountrcg commented:
mock Garmin watch commit shall be reverted when merging in dev, it is just for better testing

avouspierre commented:
I have the opportunity to have an access to vivoactive5 from my daughter. Use this feature + create the watch face for this specific watch 🔥 ➡️ 
![IMG_3827](https://github.com/user-attachments/assets/fdab1b76-f25c-40ef-a202-73609eb10c5a)


I had a issue for the update until I installed the data field (I started to install only the watch face) but not sure it

mountrcg commented:
> I had a issue for the update until I installed the data field (I started to install only the watch face) but not sure it

no the datafield / watchface are totally independant,is not needed. Just takes a minute or 2 sometimes, the g or sens arrows will disappear on the first update of watch. Up/Down also helps.

marv-out commented on Trio/Resources/json/defaults/freeaps/freeaps_settings.json:
```diff
@@ -48,5 +48,9 @@
-  "timeInRangeType": "timeInTightRange"
+  "timeInRangeType": "timeInTightRange",
+  "garminWatchface": "trio",
+  "garminDataType1": "cob",
```
there are probably more descriptive variables possible

marv-out commented on Trio/Sources/Models/Autotune.swift:
```diff
@@ -1,17 +0,0 @@
-import Foundation
```
why was there even autotune in this branch? was it based on latest dev?

mountrcg replied:
yes the files were still in dev, just the reference was removed

marv-out commented on Trio/Sources/Models/GarminWatchSettings.swift:
```diff
@@ -0,0 +1,89 @@
+
+/// Primary data type selection for Garmin watchface and datafield.
+/// Determines whether to display COB or Sensitivity Ratio alongside glucose data.
+enum GarminDataType1: String, JSON, CaseIterable, Identifiable, Codable, Hashable {
```
see comment above about naming

mountrcg replied:
### GarminWatchSettings.swift
- `GarminDataType1` (enum) → `GarminPrimaryDataField`
- `GarminDataType2` (enum) → `GarminSecondaryDataField`
- `garminDisableWatchfaceData` → `isWatchfaceDataEnabled`
  - ⚠️ **Boolean logic reversed**: Default changed from `true` (disabled) to `false` (disabled)

**New struct definition**:
```swift
struct GarminWatchSettings: Codable, Hashable {
    var watchface: GarminWatchface = .trio
    var primaryDataField: GarminPrimaryDataField = .cob
    var secondaryDataField: GarminSecondaryDataField = .tbr
    var isWatchfaceDataEnabled: Bool = false
}
```

---

### WatchConfigStateModel.swift
- `garminDataType1` → `primaryDataField`
- `garminDataType2` → `secondaryDataField`
- `garminDisableWatchfaceData` → `isWatchfaceDataEnabled`
  - ⚠️ **Boolean logic reversed**: Default changed from `true` to `false`
- `isDisableToggleLocked` → `isWatchfaceDataCooldownActive`
- `remainingCooldownSeconds` → `watchfaceSwitchCooldownSeconds`
- `cooldownTimer` → `watchfaceSwitchTimer`
- `cooldownEndTime` → `watchfaceSwitchCooldownEndTime`

avouspierre replied:
> ### GarminWatchSettings.swift
> * `GarminDataType1` (enum) → `GarminPrimaryDataField`
> * `GarminDataType2` (enum) → `GarminSecondaryDataField`
> * `garminDisableWatchfaceData` → `isWatchfaceDataEnabled`

@mountrcg  These changes of the name of the attributes must be apply in the garminWatchFace and garminFieldFace code to allow to use it. 

```
statusData = {
                "sgv" => status["sgv"],
                "..."
                "displayDataType1" => status["displayPrimaryAttributeChoice"],
                "displayDataType2" => status["displaySecondaryAttributeChoice"],
                "date" => status["date"]
            };
```

mountrcg replied:
@avouspierre I moved that around a lot. It is fixed now, but moved it to different repo, sending you invites.

marv-out commented on Trio/Sources/Services/WatchManager/GarminManager.swift:
```diff
@@ -131,36 +179,103 @@ final class BaseGarminManager: NSObject, GarminManager, Injectable {
-                        let watchState = try await self.setupGarminWatchState()
-                        let watchStateData = try JSONEncoder().encode(watchState)
-                        self.sendWatchStateData(watchStateData)
+                        // Check loop age
```
theres a lot happening now in the sink closure of the publisher. I would probably separate this into smaller parts for readability

mountrcg replied:
could you elaborate on this please, may be small example

mountrcg replied:
I haver adressed it, as I took out the different states for different watchfaces and harmonized it into one

marv-out commented on Trio/Sources/Services/WatchManager/GarminManager.swift:
```diff
@@ -131,36 +179,103 @@ final class BaseGarminManager: NSObject, GarminManager, Injectable {
-                        let watchState = try await self.setupGarminWatchState()
-                        let watchStateData = try JSONEncoder().encode(watchState)
-                        self.sendWatchStateData(watchStateData)
+                        let watchface = self.currentWatchface
```
see comment above reg. readability

marv-out commented on Trio/Sources/Services/WatchManager/GarminManager.swift:
```diff
@@ -243,21 +464,114 @@ final class BaseGarminManager: NSObject, GarminManager, Injectable {
+                throw CoreDataError.fetchError(function: #function, file: #file)
+            }
+            // Filter only events that have a tempBasal
+            return pumpEvents.filter { $0.tempBasal != nil }.map(\.objectID)
```
rather than filtering on the NSManagedObjectsArray you could use a NSCompoundPredicate and filter on db level which is much more efficient. Sth like this: 
```
        let tempBasalPredicate = NSPredicate(format: "tempBasal != nil")
        let compoundPredicate = NSCompoundPredicate(andPredicateWithSubpredicates: [
            NSPredicate.pumpHistoryLast24h,
            tempBasalPredicate
        ])
        
        let results = try await CoreDataStack.shared.fetchEntitiesAsync(
            ofType: PumpEventStored.self,
            onContext: backgroundContext,
            predicate: compoundPredicate,
            key: "timestamp",
            ascending: false, // Most recent first
            fetchLimit: 5
        )
```

mountrcg replied:
yep will try

marv-out commented on Trio/Sources/Services/WatchManager/GarminManager.swift:
```diff
@@ -327,32 +673,347 @@ final class BaseGarminManager: NSObject, GarminManager, Injectable {
+    // MARK: - SwissAlpine Watchface State Setup
+
+    /// Builds a SwissAlpine format GarminWatchState with numeric values for last 24 glucose readings
+    func setupGarminSwissAlpineWatchState() async throws -> [GarminSwissAlpineWatchState] {
```
the logic is in parts very similar to `setupGarminTrioWatchState()`. You could use helper methods in order to avoid duplicated code

mountrcg replied:
I actually ran across more issues with this. I will use only one watchstate and use the one AAPS uses, as most garmin watchface use that

mountrcg replied:
removed

marv-out commented on Trio/Sources/Services/WatchManager/GarminManager.swift:
```diff
@@ -327,32 +673,347 @@ final class BaseGarminManager: NSObject, GarminManager, Injectable {
+
+    /// Formats IOB value with one fraction digit
+    func iobFormatterWithOneFractionDigit(_ value: Decimal) -> String {
+        let formatter = NumberFormatter()
```
the formatter gets created every time the function is called. This isn't very efficient and it would be better to create it at the beginning with a static property

mountrcg replied:
redone

marv-out commented on Trio/Sources/Services/WatchManager/GarminManager.swift:
```diff
@@ -327,32 +673,347 @@ final class BaseGarminManager: NSObject, GarminManager, Injectable {
+
+    /// Formats a Date to HH:mm:ss string for logging
+    private func formatTimeForLog(_ date: Date = Date()) -> String {
+        let formatter = DateFormatter()
```
same problem with the formatter

mountrcg replied:
done

marv-out commented on Trio/Sources/Services/WatchManager/GarminManager.swift:
```diff
@@ -280,26 +594,58 @@ final class BaseGarminManager: NSObject, GarminManager, Injectable {
-                    } else {
-                        let parsedIsf = Double(truncating: insulinSensitivity).asMmolL
-                        let parsedEventualBG = Double(truncating: eventualBG).asMmolL
+                    // EventualBG with validation (stored as Int16 mg/dL in CoreData)
```
you are validating those values several times in the same way. Please use helper methods

mountrcg replied:
tried, not sure it's better

marv-out commented on Trio/Sources/Services/WatchManager/GarminManager.swift:
```diff
@@ -67,15 +68,42 @@ final class BaseGarminManager: NSObject, GarminManager, Injectable {
+
+    /// Enable simulated Garmin device for Xcode Simulator testing
+    /// When true, creates a fake Garmin device so you can test the workflow in Simulator
+    #if targetEnvironment(simulator)
```
does this work?

mountrcg replied:
yep, very nicely, it just creates a fake garmin device which allows the testing of UI elements in sim, which would not be possible without real phone and watch

marv-out commented on Trio/Sources/Services/WatchManager/GarminManager.swift:
```diff
@@ -174,65 +289,171 @@ final class BaseGarminManager: NSObject, GarminManager, Injectable {
+            return
+        }
+
+        // Start new 30s timer ONLY if none exists
```
creating a timer on the main thread is probably a bad idea and will block the UI. I would suggest using a dedicated queue

mountrcg replied:
done

marv-out commented on Trio/Sources/Services/WatchManager/GarminManager.swift:
```diff
@@ -327,32 +673,347 @@ final class BaseGarminManager: NSObject, GarminManager, Injectable {
+                }
+
+                // Process each glucose reading (up to 24)
+                for (index, glucose) in glucoseObjects.enumerated() {
```
I think this could be optimized in terms of efficiency and readability. Calculate timestamp outside of the loop, use an early return to avoid unnecessary watchstate initialization...

maybe sth like this:
```
  let mostRecentTimestamp: UInt64? = {
                    if let latestDetermination = determinationObjects.first,
                       let loopTimestamp = latestDetermination.timestamp {
                        return UInt64(loopTimestamp.timeIntervalSince1970 * 1000)
                    }
                    return nil
                }()

                // Process each glucose reading (up to 24)
                for (index, glucose) in glucoseObjects.enumerated() {
                    // Validate glucose value first
                    let glucoseValue = glucose.glucose
                    guard glucoseValue >= 0, glucoseValue <= 500 else {
                        if self.debugWatchState {
                            debug(.watchManager, "⌚️ SwissAlpine: Invalid glucose value (\(glucoseValue)), excluding from data")
                        }
                        continue
                    }

                    var watchState = GarminSwissAlpineWatchState()

                    // Set timestamp: use determination timestamp for most recent, glucose timestamp for everything previously
                    if index == 0 {
                        watchState.date = mostRecentTimestamp ?? glucose.date.map { UInt64($0.timeIntervalSince1970 * 1000) }
                    } else {
                        watchState.date = glucose.date.map { UInt64($0.timeIntervalSince1970 * 1000) }
                    }
```

mountrcg replied:
done

marv-out reviewed (changes requested):
Haven't reviewed everything (its a lot), but I already made some comments from looking at the most critical parts. I also have no Garmin device so its all only code-based

mountrcg committed:
debounce instead of throttle

clean logging

app installed check before prep

further optimizations

debounce WatchState data prep from multiple determination CD updates
add app installed check to more instances

mountrcg committed:
skip data prep when only Watchface installed but data transmission disabled

mountrcg committed:
remove cache cleaning, to skip the optimistic watchstate prep

mountrcg committed:
hash calc to enhance debounce

hasher fix

mountrcg committed:
fix throttling and datafield choice

mountrcg committed:
variable renaming

marv-out suggestion

mountrcg committed:
reduce swissalpine watchstate data

mountrcg committed:
wording to Diable Watchface data transmission

mountrcg committed:
filter at database level

marv-out suggestion

mountrcg committed:
static formatter use

marv-out suggestion

mountrcg committed:
use helper functions to validate

marv-out suggestion

mountrcg committed:
use timeQueue for debounce timer

marv-out suggestion

mountrcg committed:
optimized in terms of efficiency and readability.

marv-out suggestion

mountrcg committed:
improve debouncing multiple core data saves

more validations

mountrcg committed:
Broadcast Hash Deduplication Fix

The hash was being marked as "sent" BEFORE the actual send completed, as getAppStatus callback never fires or is severely delayed
All future broadcasts with same/similar data were blocked :
The hash from 21:40:07 is still stored in lastSentDataHash
Even though it was never actually sent via sendMessage!
So every subsequent call logs "Skipping duplicate broadcast"

mountrcg committed:
send new Glucose also immediatly if no loop within last 30min

mountrcg committed:
optimistic send

mountrcg committed:
remove answering Status requests from watch

mountrcg committed:
clean commenting

mountrcg committed:
remove obsolete code

mountrcg committed:
ConnectIQ

mountrcg committed:
remove blocking

works but also leads to cache freezes

mountrcg commented:
@Sjoerd-Bo3 & @dnzxy 
I have added 2 watchface to [connectIQ](https://apps.garmin.com/de-DE/developer/eae8d754-7a90-4cc7-85f1-95d12ea93490/apps). Can we move the PR into the feature branch to commence testing?

I moved the code for watchfaces and also 2 datafields into private repos, too much interdependance with Trio, the message queueing and energy optimization within GarminManager are too tricky.

mountrcg committed:
UUIDs cIQ store

dnzxy commented:
I‘ll merge this PR into the feature branch so it can move forward.

@marv-out if you get the chance to please give it another review, there have been quite a few changes since your last review.

mountrcg commented:
@dnzxy you mentioned in discord some changes still open from @marv-out 's input. I checked and I had addressed all.

dnzxy commented:
@mountrcg you had addressed them but they haven’t been reviewed by @marv-out :)
